be/src/olap/short_key_index.cpp

```diff
@@ -0,0 +1,153 @@
+
+#include "util/coding.h"
+#include "gutil/strings/substitute.h"
+
+using strings::Substitute;
+
+namespace doris {
+
+Status ShortKeyIndexBuilder::init() {
+    _key_length = 0;
+    for (auto& key : _key_fields) {
+        // +1 for null bit, we should improve this some day
+        _key_length += key.index_length + 1;
+    }
+    _tmp_key_buf.resize(_key_length);
+    return Status::OK();
+}
+
+Status ShortKeyIndexBuilder::add_item(const RowCursor& short_key, uint32_t block_id) {
+    char* key_buf = (char*)_tmp_key_buf.data();
+
+    size_t offset = 0;
+    for (size_t i = 0; i < _key_fields.size(); i++) {
+        short_key.write_index_by_index(i, key_buf + offset);
+        offset += short_key.get_index_size(i) + 1;
+    }
+
+    _index_buf.append(key_buf, _key_length);
+    put_fixed32_le(&_index_buf, block_id);
+
+    return Status::OK();
+}
+
+Status ShortKeyIndexBuilder::finalize(uint32_t data_length,
+                                      uint32_t num_rows,
+                                      std::vector<Slice>* Slices) {
+    // extra fixed part
+    _header.mutable_extra()->data_length = data_length;
+    _header.mutable_extra()->num_rows = num_rows;
+
+    // proto part
+    _header.mutable_message()->set_start_version(0);
+    _header.mutable_message()->set_end_version(0);
+    _header.mutable_message()->set_cumulative_version_hash(0);
+    _header.mutable_message()->set_segment(_segment_id);
+    _header.mutable_message()->set_num_rows_per_block(_num_rows_per_block);
+    _header.mutable_message()->set_delete_flag(_delete_flag);
+    _header.mutable_message()->set_null_supported(true);
+
+    std::string proto_string;
+    if (!_header.message().SerializeToString(&proto_string)) {
+        return Status::InternalError("Serialize short key header failed");
+    }
+
+    // fixed part
+    _header._fixed_file_header.magic_number = OLAP_FIX_HEADER_MAGIC_NUMBER;
+    _header._fixed_file_header.version = OLAP_DATA_VERSION_APPLIED;
+
+    _header._fixed_file_header.protobuf_length = proto_string.size();
+    _header._fixed_file_header.protobuf_checksum =
+        olap_adler32(ADLER32_INIT, proto_string.data(), proto_string.size());
+
+    // NOTE: must set file_length after protobuf_length has been set, because
+    // _header.size() depends on protobuf_length
+    _header._fixed_file_header.file_length = _index_buf.size() + _header.size();
+    _header._fixed_file_header.checksum = 
+        olap_adler32(ADLER32_INIT, (const char*)_index_buf.data(), _index_buf.size());
+
+    _header_buf.resize(_header.size());
+    char* ptr = (char*)_header_buf.data();
+
+    memcpy(ptr, &_header._fixed_file_header, sizeof(_header._fixed_file_header));
+    ptr += sizeof(_header._fixed_file_header);
+
+    memcpy(ptr, &_header._extra_fixed_header, sizeof(_header._extra_fixed_header));
+    ptr += sizeof(_header._extra_fixed_header);
+
+    memcpy(ptr, proto_string.data(), proto_string.size());
+
+    Slices->emplace_back(_header_buf);
```

Comment:
here can also determine whether to return the header buf depending on Whether it is V1 or V2?

Reply:
Same with block_id, we do this after smoke test